Profiler/Basics/BaseProfiler.cs

```diff
@@ -0,0 +1,94 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using NLog;
+using Profiler.Core;
+using TorchUtils;
+
+namespace Profiler.Basics
+{
+    /// <summary>
+    /// Implement basic features of a profiler that makes use of ProfilerEntry.
+    /// </summary>
+    /// <remarks>You can use ProfilerPatch without this class.</remarks>
+    public abstract class BaseProfiler<K> : IProfiler, IDisposable
+    {
+        // Thread-safe dictionary of ProfilerEntry with an arbitrary type of keys.
+        readonly ConcurrentDictionary<K, ProfilerEntry> _profilerEntries;
+
+        // Cached function to unpool (or create) a new ProfilerEntity instance.
+        readonly Func<K, ProfilerEntry> _makeProfilerEntity;
+
+        ulong _startFrameCount;
+        DateTime _startTime;
+
+        protected BaseProfiler()
+        {
+            _profilerEntries = new ConcurrentDictionary<K, ProfilerEntry>();
+            _makeProfilerEntity = _ => ProfilerEntry.Pool.Instance.UnpoolOrCreate();
+        }
+
+        /// <summary>
+        /// Mark the beginning of profiling. Must be called once to properly retrieve the profiling data.
+        /// </summary>
+        public void MarkStart()
+        {
+            _startFrameCount = VRageUtils.CurrentGameFrameCount;
+            _startTime = DateTime.UtcNow;
+        }
+
+        /// <inheritdoc/>
+        void IProfiler.ReceiveProfilerResult(in ProfilerResult profilerResult)
+        {
+            try
+            {
+                if (TryAccept(profilerResult, out var key))
+                {
+                    var profilerEntry = _profilerEntries.GetOrAdd(key, _makeProfilerEntity);
```

Comment:
This is always called from the same thread, so I don't think you need a concurrent dictionary here.

Reply:
Commands in `ProfilerCommands` do consume the dictionary contents from different threads (eg. `GetResults()` after `Task.Delay()`) so I'd keep it a concurrent dictionary for now.